File: ex3/ex3.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> charset;
        int left = 0, right = 0;
        int len = s.size();
        int maxlen = 0, currlen = 0;
        while (right < len)
        {
            if (charset.count(s[right]))
            {
                left = charset[s[right]] + 1;
                for (auto i = charset.begin(); i != charset.end();)
                {
                    if (i->second < left)
                    {
                        auto tmp = i;
                        ++i;
                        charset.erase(tmp->first);
                    }
                    else
                        ++i;
                }
            }
            charset[s[right]] = right;
            currlen = right - left + 1;
            maxlen = max(currlen, maxlen);
            right++;
        }
        return maxlen;
    }
};

int main(void)
{
    Solution sy;
    cout << sy.lengthOfLongestSubstring("abba") << endl;
    return 0;
}

File: ex3/ex3.py
#!/usr/bin/env python3

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        charset = set()
        l = r = 0
        maxlen = 0
        while r < len(s):
            if s[r] not in charset:
                charset.add(s[r])
                maxlen = max(maxlen, r - l + 1)
                r += 1
            else:
                charset.remove(s[l])
                l += 1
        return maxlen

sy = Solution()
print(sy.lengthOfLongestSubstring("abba"))